src/shm/pool/raw.rs

```diff
@@ -0,0 +1,235 @@
+//! A raw shared memory pool handler.
+//!
+//! This is intended as a safe building block for higher level shared memory pool abstractions and is not
+//! encouraged for most library users.
+
+use std::{
+    fs::File,
+    io,
+    os::unix::prelude::{FromRawFd, RawFd},
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use memmap2::MmapMut;
+use nix::{
+    errno::Errno,
+    fcntl,
+    sys::{mman, stat},
+    unistd,
+};
+use wayland_client::{
+    backend::InvalidId,
+    protocol::{wl_buffer, wl_shm, wl_shm_pool},
+    ConnectionHandle, Dispatch, QueueHandle,
+};
+
+use super::CreatePoolError;
+
+/// A raw handler for file backed shared memory pools.
+///
+/// This type of pool will create the SHM memory pool and provide a way to resize the pool.
+///
+/// This pool does not release of buffers. If you need this, use one of the higher level pools.
+#[derive(Debug)]
+pub struct RawPool {
+    pool: wl_shm_pool::WlShmPool,
+    len: usize,
+    mem_file: File,
+    mmap: MmapMut,
+}
+
+impl RawPool {
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize, cx: &mut ConnectionHandle) -> io::Result<()> {
+        if size > self.len {
+            self.len = size;
+            self.mem_file.set_len(size as u64)?;
+            self.pool.resize(cx, size as i32);
+            self.mmap = unsafe { MmapMut::map_mut(&self.mem_file) }?;
+        }
+
+        Ok(())
+    }
+
+    /// Returns a reference to the underlying shared memory file using the memmap2 crate.
+    pub fn mmap(&mut self) -> &mut MmapMut {
+        &mut self.mmap
+    }
+
+    /// Create a new buffer to this pool.
+    ///
+    /// ## Parameters
+    /// - `offset`: the offset (in bytes) from the beginning of the pool at which this buffer starts.
+    /// - `width` and `height`: the width and height of the buffer in pixels.
+    /// - `stride`: distance (in bytes) between the beginning of a row and the next one.
+    /// - `format`: the encoding format of the pixels.
+    ///
+    /// The encoding format of the pixels must be supported by the compositor or else a protocol error is
+    /// risen. You can ensure the format is supported by listening to [`ShmHandler::supported_format`](super::ShmHandler::supported_format).
+    #[allow(clippy::too_many_arguments)]
+    pub fn create_buffer<D, U>(
+        &mut self,
+        offset: i32,
+        width: i32,
+        height: i32,
+        stride: i32,
+        format: wl_shm::Format,
+        udata: U,
+        cx: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+    ) -> Result<(&mut [u8], wl_buffer::WlBuffer), InvalidId>
```

Comment:
I like this `RawPool` abstraction overall, but I'm not sure the `create_buffer` method should return a `&mut [u8]` slice to the underlying buffer. It's not necessary  for a low-level "raw" pool, and I'm afraid it will not interact well with higher-level constructs, and most notably buffer-reuse.